Publish a dual CJS/ESM package with platform-specific loaders

sandhose opened the pull request:
This package currently ships the WASM module as a JS string, base64-encoded.

This is very inefficient for three reasons:

 - the base64 encoding is about 30% larger than the WASM module
 - it decodes on the JS side, which means dealing with a large string in JS memory, and hanging the main thread for a long time
 - it means we can't use the streaming compilation APIs

For it to be handled correctly by bundlers, it is better to provide an ESM-compatible package. Because some environments (like the JS SDK test suite) still run on CJS, I've made it a proper dual CJS/ESM package.

The loading logic is also platform-specific: on node-like environments, it needs to read the file from the disk, whereas on browser-like environments it needs to `fetch` the WASM file. This is why this now provides platform-specific entrypoints.

To give an idea of what performance impact this change has, here is a before/after comparison on Element Call:

<img width="1283" alt="image" src="https://github.com/user-attachments/assets/be22eae1-b99c-430d-943e-e69796e7b50e">

Notice how:

 - the WASM is a lot smaller to download than the JS blob
 - it compiles during the download
 - there is a big long task on the 'before' which is just the base64-decoding. 660ms on that profiling, with a 4x CPU slowdown
 - it saves around 10MB in the JS heap memory

Same before/after on Element Web:

![image](https://github.com/user-attachments/assets/d20e2a0d-d4df-4508-a3b2-7a1e7007af7f)

sandhose commented on .eslintrc.cjs:
This is because of the 'type: module' change in the package.json

sandhose commented on index.js:
Instead of wrangling with the output of wasm-pack, we instead create two new entrypoints, one for node, the other for other platforms (which support fetch, so mainly browsers)

sandhose commented on src/lib.rs:
```diff
@@ -84,3 +84,6 @@ pub fn get_versions() -> Versions {
 }
+
+#[cfg(test)]
+wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_node_experimental);
```
This is to enable ES modules in node in the wasm_bindgen test environment. Else it won't work as we switched the package to be a module

richvdh replied:
Do we still need this, now that the package is dual commonjs/esm?

sandhose replied:
It *was* necessary because I've flipped the main module type before to ESM (`"type": "module"` in the package.json) which made it that `.js` files were considered as ESM not CJS, which broke the regular test.

sandhose replied:
This was reverted

sandhose commented on tests/device.test.js:
```diff
@@ -229,11 +242,19 @@ describe("Send to-device message", () => {
 describe("Key Verification", () => {
-    const userId1 = new UserId("@alice:example.org");
-    const deviceId1 = new DeviceId("alice_device");
+    let userId1;
+    let deviceId1;
```
Those kind of changes are because we don't load the WASM module synchronously anymore, so accessing them before it's loaded (so before the `beforeAll` call in the `helper.js`)

Hywan replied:
👍

richvdh replied:
Are these changes still necessary, now that the package works if you try to use it synchronously (at least in nodejs) ?

sandhose commented on tests/tracing.test.js:
Removed those tests, as it looks like tracing is always enabled nowadays

richvdh commented:
As discussed: unfortunately, this is going to break the Jest tests in matrix-js-sdk. We looked into trying to get Jest running on ES modules, but ultimately [gave up](https://github.com/matrix-org/matrix-js-sdk/pull/4187#issuecomment-2117342908).

Hywan commented on tests/helper.js:
```diff
@@ -1,6 +1,10 @@
+import { initAsync } from "../";
 
-function* zip(...arrays) {
+beforeAll(() => initAsync());
```
Why this here?

Hywan replied:
Oh, it's because `tests/helper.js` is defined as a setup file in vitest config. Maybe a comment would be welcomed.

sandhose replied:
This was removed

Hywan commented on tests/tracing.test.js:
Can't we migrate these tests?

sandhose replied:
You're right, I don't know why it wasn't working in my first attempt. I've restored it

sandhose commented:
I switched gears a little bit, and instead opted for making the package have 4 entrypoints:

 - Node.js with CJS (`node.cjs`)
 - Node.js with ESM (`node.js`)
 - browser with CJS (`index.cjs`)
 - browser with ESM (`index.js`)

One annoying thing, is that the matrix-js-sdk expects the WASM to be already loaded and available, and doesn't call `initAsync`, so I made the Node.js versions load synchronously.

To achieve that, I used the `bundler` target of `wasm-pack`, and removed their loading logic for my own.

richvdh commented on index.cjs:
```diff
@@ -0,0 +1,34 @@
+const wasm = require("./pkg/matrix_sdk_crypto_wasm_bg.cjs");
```
Again, I find it confusing to call this `wasm`.

richvdh commented on index.cjs:
```diff
@@ -0,0 +1,34 @@
+}
+
+module.exports = {
+    // Re-export everything from the wasm module bindings
```
```suggestion
    // Re-export everything from the generated javascript wrappers
```

richvdh commented on index.js:
```diff
@@ -0,0 +1,31 @@
+export * from "./pkg/matrix_sdk_crypto_wasm_bg.js";
+import * as wasm from "./pkg/matrix_sdk_crypto_wasm_bg.js";
+
+const moduleUrl = new URL("./pkg/matrix_sdk_crypto_wasm_bg.wasm", import.meta.url);
```
for reasons I don't entirely understand, webpack is putting the wasm in the top-level build directory, rather than in a `bundle` directory. That's probably a problem in element-web's webpack config, but it would be good to fix before we ship this.

sandhose replied:
I also noticed that, no idea why. Element Call (so, Vite) properly handled this as regular asset.

richvdh reviewed (changes requested):
A few questions and comments.

This is a bit of a scary-looking PR. Do you think it would be possible to break it up a bit? (For instance, could we switch to vitest in a separate PR?)

sandhose commented on scripts/build.sh:
```diff
@@ -1,48 +1,26 @@
-cd $(dirname "$0")/..
+cd "$(dirname "$0")"/..
 
-# --no-pack disables generation of a `package.json` file. Such a file is
-# useless to us (our package uses the hand-written one in the root directory),
-# and contains incorrect data (our `main` is `index.js`).
```
I've restored it, reworded a little bit

sandhose commented on scripts/build.sh:
```diff
@@ -1,48 +1,26 @@
-#  2. Remove the imports of `TextDecoder` and `TextEncoder`. We rely on the global defaults.
+# This will output two important files:
+#   - pkg/matrix_sdk_crypto_wasm.js
+#   - pkg/matrix_sdk_crypto_wasm_bg.js
```
I've expanded this comment to list all the files

sandhose commented on scripts/build.sh:
```diff
@@ -1,48 +1,26 @@
-  cat scripts/epilogue.js
-} > pkg/index.js
+# We're only interested in the last one, as the first one is the loader,
+# and the output from wasm-bindgen doesn't work well on all platforms, so we ship our own loader.
```
I did that

sandhose commented on scripts/build.sh:
```diff
@@ -1,48 +1,26 @@
+# We're only interested in the last one, as the first one is the loader,
+# and the output from wasm-bindgen doesn't work well on all platforms, so we ship our own loader.
+# We still want to convert the last one to  CommonJS, so we ship a proper dual commonjs/es6 module.
 
```
I did something similar

sandhose commented on node.cjs:
```diff
@@ -0,0 +1,24 @@
+const filename = path.join(__dirname, "pkg/matrix_sdk_crypto_wasm_bg.wasm");
+
+// In Node environments, we need to load the WASM module synchronously, as
+// the test suite in matrix-js-sdk don't call `initAsync`
```
So, it was annoying to restore this behaviour with both CommonJS and ESM compatibility; especially the latter, where it's impossible to export a `Proxy` like we basically did before.

I instead opted for another solution: add a second entrypoint which doesn't export anything, but loads the module synchronously. You'd basically use it like that:


```js
require("@matrix-org/matrix-sdk-crypto-wasm/init-sync");
const { UserID } = require("@matrix-org/matrix-sdk-crypto-wasm");

const u = new UserID("@alice:example.com");
```

or in ESM environments:

```js
import "@matrix-org/matrix-sdk-crypto-wasm/init-sync";
import { UserID } from "@matrix-org/matrix-sdk-crypto-wasm";

const u = new UserID("@alice:example.com");
```

The sync loader would only work in Node though (or any environment that has `fs.readFileSync` polyfilled) but I feel like this is fine? The alternative would be for us to ship a `export const wasm = "…base64 encoded thing…";` again, which I don't think we want to do.

For the JS SDK, this means that we'll need to add this import in the tests using it, *or* make the tests call `loadAsync` somehow

sandhose replied:
The synchronous loading behaviour was restored

sandhose commented on index.cjs:
```diff
@@ -0,0 +1,34 @@
+        });
+        wasm.__wbg_set_wasm(instance.exports);
+        instance.exports.__wbindgen_start();
+    },
```
See the comment on `node.cjs`. Noted on the changelog entry, will do if we agree on this new solution

sandhose replied:
The behaviour was restored

sandhose commented on index.js:
```diff
@@ -0,0 +1,31 @@
+export * from "./pkg/matrix_sdk_crypto_wasm_bg.js";
+import * as wasm from "./pkg/matrix_sdk_crypto_wasm_bg.js";
```
I've called it `bindings` now everywhere

sandhose commented on index.cjs:
I've done that

sandhose commented:
@richvdh I've reverted the move to vitest, now that it works correctly again in CJS environments 💃 

For consumers, the bindings don't synchronously initialise themselves automatically anymore. If they need that behaviour, they need to import `@matrix-org/matrix-sdk-crypto-wasm/init-sync`. I haven't updated the README/changelog yet, I first want feedback on this solution

richvdh commented on index.js:
```diff
@@ -0,0 +1,51 @@
+    const instance = new WebAssembly.Instance(module, {
+        "./matrix_sdk_crypto_wasm_bg.js": bindings,
+    });
+    bindings.bindings(instance.exports);
```
shouldn't this be `bindings.__wbg_set_wasm(instance.exports);` ?

sandhose replied:
It should, not sure how it ended up here. Fixed.

richvdh commented on index.mjs:
```diff
@@ -0,0 +1,49 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This is the entrypoint on non-node ESM environments.
```
```suggestion
// This is the entrypoint on non-node ESM environments (such as Element Web).
```

sandhose commented on index.mjs:
```diff
@@ -0,0 +1,49 @@
+    if (!response.ok) {
+        throw new Error(`Failed to fetch wasm module: ${moduleUrl}`);
+    }
+    mod = await WebAssembly.compile(response);
```
(I miss typechecking), you're right we should call response.arrayBuffer() first.

sandhose replied:
This was fixed & I added typechecking

sandhose commented:
@richvdh I've added typechecking of JS files to avoid some of the dumb errors you've spotted previously

richvdh commented on tsconfig.json:
```diff
@@ -1,6 +1,7 @@
+        "lib": ["DOM"]
     },
     "typedocOptions": {
         "entryPoints": ["pkg/matrix_sdk_crypto_wasm.d.ts"],
```
this presumably needs an update, to make sure that initAsync continues to be documented?

sandhose replied:
Good point, fixed in [b89ec58](https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/pull/167/commits/b89ec58f9b767d047fa3312a72c2382ca6995641)

richvdh reviewed (changes requested):
This looks like a sensible direction to me now. Some general points:

* I've pulled out the test file changes into a separate PR (#171). Could you rebase on top of that?
* There's a lot of duplicate code between the entry points. I'd like to see if we can factor a lot of that out to one or two shared modules (possibly: one module which is shared across all four entry points and one which is shared between the two `node` entry points.) AIUI a commonjs module can be easily `import`ed into an ESM module?
* I'd like to see comments on the various function definitions describing what they do, which global variables they touch, what they return, etc.

richvdh commented:
A few other things I meant to ask you to do but forgot:
 * Please can you make sure the README is up-to-date (it may well be fine, but please double-check)
 * Please can you stick something in the changelog about this
 * Could you go through the PR and mark any significantly-outdated comments as resolved?

sandhose commented:
@richvdh 

> I've pulled out the test file changes into a separate PR (https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/pull/171). Could you rebase on top of that?

Done

> There's a lot of duplicate code between the entry points. I'd like to see if we can factor a lot of that out to one or two shared modules (possibly: one module which is shared across all four entry points and one which is shared between the two node entry points.) AIUI a commonjs module can be easily imported into an ESM module?

I'm really not sure about sharing code. I would like to avoid crossing CJS/ESM boundaries, it's well-defined enough IMO

> I'd like to see comments on the various function definitions describing what they do, which global variables they touch, what they return, etc.

Done, the JSDoc annotations helped me spot a bad return 👍 

> Please can you make sure the README is up-to-date (it may well be fine, but please double-check)

Done

> Please can you stick something in the changelog about this

Done, shouted about the breaking change of it not synchronously initialising in Web environments anymore

> Could you go through the PR and mark any significantly-outdated comments as resolved?

I've gone through all of them I think

richvdh commented on index.js:
```diff
@@ -0,0 +1,97 @@
+
+// @ts-check
+
+// This is the entrypoint on non-node CommonJS environments.
+// `asyncLoad` will load the WASM module using a `fetch` call.
+const bindings = require("./pkg/matrix_sdk_crypto_wasm_bg.cjs");
```
(likewise in the other entry points)

sandhose replied:
Done in 7e7c15c

richvdh commented on node.js:
```diff
@@ -0,0 +1,114 @@
+
+const filename = path.join(__dirname, "pkg/matrix_sdk_crypto_wasm_bg.wasm");
+
+// We want to automatically load the WASM module in Node environments
+// synchronously if the consumer did not call `initAsync`. To do so, we install
+// a `Proxy` that will intercept calls to the WASM module
```
(likewise in node.mjs)

sandhose replied:
Done in 7e7c15c

richvdh reviewed (commented):
Sorry this is taking so many iterations. I think it's nearly there though.

sandhose commented on index.js:
```diff
@@ -0,0 +1,97 @@
+
+    if (initialised) return;
+    initialised = true;
+    /** @type {{exports: typeof import("./pkg/matrix_sdk_crypto_wasm_bg.wasm.d")}} */
+    // @ts-expect-error: Typescript doesn't know what the instance exports exactly
+    const instance = new WebAssembly.Instance(mod, {
+        // @ts-expect-error: The bindings don't exactly match the 'ExportValue' type
+        "./matrix_sdk_crypto_wasm_bg.js": bindings,
+    });
+
+    bindings.__wbg_set_wasm(instance.exports);
+    instance.exports.__wbindgen_start();
```
Right, it just means we can't do the early return for the `compileStreaming` but it's less code

sandhose commented on README.md:
```diff
@@ -26,27 +26,25 @@ Encryption](https://en.wikipedia.org/wiki/End-to-end_encryption)) for
-
-    It is recommended that you use a dynamic import, particularly in a Web
-    environment, because the WASM artifiact is large:
+   The library must be initialised before it can be used, else it will throw an error.
```
Done in 7202b78

sandhose commented on README.md:
```diff
@@ -26,27 +26,25 @@ Encryption](https://en.wikipedia.org/wiki/End-to-end_encryption)) for
-    It is recommended that you use a dynamic import, particularly in a Web
-    environment, because the WASM artifiact is large:
+   The library must be initialised before it can be used, else it will throw an error.
+   On Web platforms, the library will use `fetch` to get the WebAssembly binary, which should be handled by most bundlers.
```
Done in 7202b78

sandhose commented on index.mjs:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @returns {Promise<void>}
+ */
+async function initAsync() {
```
Forgot to push this morning 😮‍💨 
Done in 0a596f8

sandhose commented on node.js:
```diff
@@ -0,0 +1,114 @@
+async function initAsync() {
+    if (!modPromise) modPromise = loadModule();
+    const mod = await modPromise;
+    if (initialised) return;
+    initialised = true;
```
We still need to return the exports in initInstance to be able to return it in the Proxy, which we won't be able to do if we do an early return, unless we store it.

Regardless, `loadModuleSync` will throw is `modPromise` is set, which is done as soon as `initAsync` is called, and it's all sync, so no possible race condition between await points

richvdh replied:
> Regardless, loadModuleSync will throw is modPromise is set, which is done as soon as initAsync is called, and it's all sync, so no possible race condition between await points

My concern is people calling `initAsync` *after* using one of the entry points locally, (and hence hitting `loadModuleSync`).

The synchronous load path does not set `modPromise` or `initialised`, so a call to `initAsync` will fetch, compile, and re-initialise the module, meaning we have two copies of the wasm module and a big mess.

> We still need to return the exports in `initInstance` to be able to return it in the Proxy, which we won't be able to do if we do an early return, unless we store it.

I think you misread my suggestion. Will make it more explicit.

sandhose replied:
Riiiight, got it. Implemented that in d7d3424

sandhose commented on index.js:
```diff
@@ -0,0 +1,97 @@
+let modPromise = null;
+
+/**
+ * Tracks whether the module has been instanciated or not
```
Done in 7e7c15c

richvdh commented on node.js:
```diff
@@ -0,0 +1,117 @@
+ * @param {WebAssembly.Module} mod
+ * @returns {typeof import("./pkg/matrix_sdk_crypto_wasm_bg.wasm.d")}
+ */
+function initInstance(mod) {
```
```suggestion
function initInstance(mod) {
    if (initialised) {
        // This should be unreachable
        throw new Error("initInstance called twice");
    }
    
```

sandhose replied:
This actually throws if you call `initAsync`, it loads in the background, and then init it synchronously before it finishes; but I think it is fine to throw because the consumer shouldn't do that

richvdh replied:
> This actually throws if you call initAsync, it loads in the background, and then init it synchronously before it finishes;

Will it? In that case, wouldn't `modPromise` be set, so that the synchronous init will fail early with an exception?

richvdh commented on node.js:
```diff
@@ -0,0 +1,117 @@
+    });
+
+    bindings.__wbg_set_wasm(instance.exports);
+    instance.exports.__wbindgen_start();
```
```suggestion
    instance.exports.__wbindgen_start();
    initialised = true;
```

richvdh commented on node.js:
```diff
@@ -0,0 +1,117 @@
+ *
+ * @returns {Promise<void>}
+ */
+async function initAsync() {
+    if (!modPromise) modPromise = loadModule();
+    const mod = await modPromise;
+    if (initialised) return;
+    initialised = true;
+    initInstance(mod);
+}
```
```suggestion
async function initAsync() {
    if (initialised) return;
    if (!modPromise) {
        modPromise = loadModule().then((mod) => initinstance(mod));
    }
    await modPromise;
}
```

richvdh reviewed (approved):
lgtm! Thanks for doing this!

richvdh commented on CHANGELOG.md:
```diff
@@ -1,6 +1,16 @@
 -   Update matrix-rusk-sdk to `e99939db857ca`.
+-   The published package is now a proper dual CommonJS/ESM package.
+-   The WebAssembly module is now loaded using `fetch` on Web platforms, reducing
+    the bundle size significantly, as well as the time it takes to compile it.
```
```suggestion
    the bundle size significantly, as well as the time it takes to compile it.
    ([#167](https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/pull/167))
```
